src/map.cpp

```diff
@@ -529,62 +529,146 @@ bool Map::canThrowObjectTo(const Position& fromPos, const Position& toPos, bool
+{
+	float dx = x1 - x0;
+	float dy = y1 - y0;
+	float grad = (dx == 0) ? 1 : dy / dx;
+	float xy = y0 + grad;
+
+	for (uint16_t y = x0 + 1; y < x1; ++y) {
+		uint16_t newX = std::floor(xy);
+		bool tileClear = isTileClear(newX, y, z);
+		if (!tileClear) {
+			return false;
 		}
+		xy += grad;
+	}
 
-		if (start.x != destination.x && (start.y == destination.y || move_ver > move_hor || move_ver > move_cross)) {
-			start.x += mx;
-		}
+	return true;
+}
 
-		const Tile* tile = getTile(start.x, start.y, start.z);
-		if (tile && tile->hasProperty(CONST_PROP_BLOCKPROJECTILE)) {
+bool Map::getSlightLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	float dx = x1 - x0;
+	float dy = y1 - y0;
+	float grad = (dx == 0) ? 1 : dy / dx;
+	float xy = y0 + grad;
+
+	for (uint16_t x = x0 + 1; x < x1; ++x) {
+		uint16_t newY = std::floor(xy + 0.1);
+		bool tileClear = isTileClear(x, newY, z);
+		if (!tileClear) {
 			return false;
 		}
+		xy += grad;
 	}
 
-	// now we need to perform a jump between floors to see if everything is clear (literally)
-	while (start.z != destination.z) {
-		const Tile* tile = getTile(start.x, start.y, start.z);
-		if (tile && tile->getThingCount() > 0) {
-			return false;
+	return true;
+}
+
+bool Map::checkSightLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	if (x0 == x1 && y0 == y1) {
+		return true;
+	}
+
+	if (std::abs(y1 - y0) > std::abs(x1 - x0)) {
+		if (y1 > y0) {
+			return getSteepLine(y0, x0, y1, x1, z);
 		}
+		return getSteepLine(y1, x1, y0, x0, z);
+	}
 
-		start.z++;
+	if (x0 > x1) {
+		return getSlightLine(x1, y1, x0, y0, z);
 	}
 
-	return true;
+	return getSlightLine(x0, y0, x1, y1, z);
 }
 
-bool Map::isSightClear(const Position& fromPos, const Position& toPos, bool floorCheck) const
+bool Map::isSightClear(const Position& fromPos, const Position& toPos, bool floorCheck /*= false*/) const
 {
-	if (floorCheck && fromPos.z != toPos.z) {
-		return false;
-	}
+	uint16_t x0 = fromPos.x;
+	uint16_t y0 = fromPos.y;
+	uint8_t z0 = fromPos.z;
+
+	uint16_t x1 = toPos.x;
+	uint16_t y1 = toPos.y;
+	uint8_t z1 = toPos.z;
 
-	// Cast two converging rays and see if either yields a result.
-	return checkSightLine(fromPos, toPos) || checkSightLine(toPos, fromPos);
+	if (z0 == z1) {
+		//target is on the same floor
+
+		//skip checks if toPos is next to us
+		if (Position::getDistanceX(fromPos, toPos) < 2 && Position::getDistanceY(fromPos, toPos) < 2) {
+			return true;
+		}
+
+		//sight is clear or floorCheck is enabled
+		bool sightClear = checkSightLine(x0, y0, x1, y1, z0);
+		if (sightClear || floorCheck) {
+			return sightClear;
+		}
+
+		//no obstacles above floor 0 so we can throw above the obstacle
+		if (z0 == 0) {
+			return true;
+		}
+
+		//check if tiles above us and target are clear and check for clear sight between them
+		--z0;
+		return checkSightLine(x0, y0, x1, y1, z0) && isTileClear(x0, y0, z0, true) && isTileClear(x1, y1, z0, true);
+	} else {
```

Comment:
this else is pretty much redundant if we gonna return in first if anyway, so I would remove it and just continue the code